compiler/rustc_middle/src/ty/util.rs

```diff
@@ -828,6 +832,61 @@ impl<'tcx> ty::TyS<'tcx> {
         }
     }
 
+    pub fn gc_layout(&'tcx self, tcx: TyCtxt<'tcx>, param_env: ty::ParamEnv<'tcx>) -> (u64, u64) {
+        let layout = tcx.layout_of(param_env.and(self)).unwrap();
+        let w_size = Size::from_bytes(size_of::<usize>());
+        let w_align = Align::from_bytes(w_size.bytes()).unwrap();
+
+        if layout.align.abi.bytes() < (w_size.bytes())
+            || self.is_no_trace(tcx.at(DUMMY_SP), param_env)
+        {
+            // If an ADT is aligned to a size smaller than 1 word, then it
+            // contains no fields large enough to represent a GC pointer.
+            //
+            // By default, the collector will conservatively scan a struct's
+            // memory, treating it as word-sized chunks. However, since rust is
+            // free to shuffle field order around, any in-memory words which
+            // represent actual pointers would be purely accidental. Such
+            // structs should be ignored completely by the collector.
+            return (0, 0);
+        }
+
+        let src_field_iter = self.ty_adt_def().unwrap().all_fields();
+        let mut bitmap: u64 = u64::MAX;
+
+        for i in layout.fields.index_by_increasing_offset() {
+            let target_offset = layout.fields.offset(usize::from(i));
+
+            let layout_cx = LayoutCx { tcx, param_env };
+            let field = layout.field(&layout_cx, i).unwrap();
+
+            if field.size < w_size {
+                // If a field is smaller than a word, it will either be shuffled
+                // around in the struct or padded to the min alignment with
+                // uninitialized memory. Whatever the case, the word it lives in is
+                // never going to be a valid pointer.
+                //
+                // target_offset is the byte in the struct's layout where this field
+                // begins. Diving by w_align gives the offset in words.
+                let mask = target_offset.bytes() / w_align.bytes();
+                bitmap &= !(1 << mask);
+            }
+
+            let no_trace = field.ty.is_no_trace(tcx.at(DUMMY_SP), param_env);
+
+            if no_trace || tcx.has_attr(src_field_iter.clone().nth(i).unwrap().did, sym::no_trace) {
+                let start = target_offset.bytes() / w_align.bytes();
+                let end = start + (field.size.bytes() / w_align.bytes());
+                for n in start..end {
```

Comment:
I thought "there must be a more elegant way to do this". But it's actually quite hard it seems https://stackoverflow.com/questions/45352352/most-efficient-way-to-set-n-consecutive-bits-to-1! [Rust does have some sort of support bzhi, but I can't see that it's wrapped as part of a nicer and portable function.]

Reply:
Yes, I didn't particularly like this either!